rebuild-rels.php

```diff
@@ -0,0 +1,203 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Fig\Link\Builder;
+
+// Download this file from https://www.iana.org/assignments/link-relations/link-relations.xml
+// It cannot be auto-downloaded by the script because IANA has scripts blocked from accessing it.
+// I cannot fathom how that makes any sense whatsoever.
+const REGISTRY_FILE = 'link-relations.xml';
+
+run();
+
+function run() {
+    $compiler = new RegistryCompiler();
+
+    $out = fopen('src/Relations.php', 'w');
+    $compiler->compile(getRecords(), $out);
+}
+
+/**
+ * Fetches the records for the relation registry.
+ *
+ * @return iterable<\SimpleXMLElement>
+ */
+function getRecords()
+{
+    $registry = simplexml_load_file(REGISTRY_FILE);
+
+    foreach ($registry->registry->children() as $element) {
+        if ($element->getName() == 'record') {
+            yield $element;
+        }
+    }
+}
+
+/**
+ * Compiles a list of SimpleXmlElements to a Relations interface.
+ */
+class RegistryCompiler
+{
+    /**
+     * Writes a Relations index class based on the provided records.
+     *
+     * @param iterable<\SimpleXMLElement> $records
+     *   An iterable of SimpleXml elements from the Link Relations XML file.
+     * @param $stream
+     *   An open file stream to which to write the generated code.
+     * @param string $class
+     *   The name of the interface to produce.
+     * @param string $namespace
+     *   The namespace of the interface to generate.
+     */
+    public function compile(iterable $records, $stream, string $class = 'Relations', string $namespace = 'Fig\\Link') : void
+    {
+        fwrite($stream, $this->createPreamble($class, $namespace));
+
+        foreach ($records as $record) {
+            $item = $this->createEntry($record);
+            fwrite($stream, $item);
+        }
+
+        fwrite($stream, $this->createClosing());
+    }
+
+    /**
+     * Processes a SimpleXml record into a constant definition.
+     *
+     * @param \SimpleXMLElement $record
+     *   The record to process.
+     * @return string
+     */
+    protected function createEntry(\SimpleXMLElement $record) : string
+    {
+        $value = (string)$record->value;
+        $name = $this->buildName($value);
+        $description = $this->rewrap((string)$record->description);
+        $note = $this->rewrap((string)$record->note);
+        $seeUri = $this->xrefLink($record->spec->xref);
+
+        return <<<END
+
+    /**
+     * {$description}
+     *
+     * {$note}
```

Comment:
The majority of the entries do not have notes, which means we end up with the majority having structures like this:

```php
/**
  * Refers to...
  *
  *
  *
  * @see iana link
  */
```

In other words, two extra lines for each docblock.

I'm thinking that maybe this should be:

```php
/**
  * {$description}{$note}
  *
  * @see {$seeUri}
  */
```

If a note is detected, then it uses the following format:

```php
private const NOTE_TEMPLATE = <<< 'END'

 * {$note}
 *
END;
```

Yes, a bit more work, but it makes the source more readable in the end.